software/contrib/euclidean_rhythm.py

```diff
@@ -0,0 +1,356 @@
+                else:
+                    if self.notes_para[self.selection_idx][1] < 0: self.notes_para[self.selection_idx][1] = 0
+                if self.notes_para[self.selection_idx][2] > self.notes_para[self.selection_idx][0]: self.notes_para[self.selection_idx][2] = self.notes_para[self.selection_idx][0]
+                if self.notes_para[self.selection_idx][2] < 0: self.notes_para[self.selection_idx][2] = 0
+                if self.notes_para[self.selection_idx][3] > 3: self.notes_para[self.selection_idx][3] = 3
+                if self.notes_para[self.selection_idx][3] < 0: self.notes_para[self.selection_idx][3] = 0
+                #print(self.selection_idx, self.notes_para[self.selection_idx])
+            else:
+                self.clock_flg = 0
+        
+        @b2.handler_falling
+        def para_add():
+            if self.selection_idx <= 5:
+                if self.editing_idx == 0:
+                    self.notes_para[self.selection_idx][0] += 1 # length
+                elif self.editing_idx == 1:
+                    self.notes_para[self.selection_idx][1] += 1 # fill
+                elif self.editing_idx == 2:
+                    self.notes_para[self.selection_idx][2] += 1 # offset
+                elif self.editing_idx == 3:
+                    self.notes_para[self.selection_idx][3] += 1 # CV
+                
+                # Bound
+                if self.notes_para[self.selection_idx][0] > 19: self.notes_para[self.selection_idx][0] = 19
+                if self.notes_para[self.selection_idx][0] < 1: self.notes_para[self.selection_idx][0] = 1
+                if self.notes_para[self.selection_idx][1] > self.notes_para[self.selection_idx][0]: self.notes_para[self.selection_idx][1] = self.notes_para[self.selection_idx][0]
+                if self.notes_para[self.selection_idx][0] > 1:
+                    if self.notes_para[self.selection_idx][1] < 1: self.notes_para[self.selection_idx][1] = 1
+                else:
+                    if self.notes_para[self.selection_idx][1] < 0: self.notes_para[self.selection_idx][1] = 0
+                if self.notes_para[self.selection_idx][2] > self.notes_para[self.selection_idx][0]: self.notes_para[self.selection_idx][2] = self.notes_para[self.selection_idx][0]
+                if self.notes_para[self.selection_idx][2] < 0: self.notes_para[self.selection_idx][2] = 0
+                if self.notes_para[self.selection_idx][3] > 3: self.notes_para[self.selection_idx][3] = 3
+                if self.notes_para[self.selection_idx][3] < 0: self.notes_para[self.selection_idx][3] = 0
+                #print(self.selection_idx, self.notes_para[self.selection_idx])
+            else:
+                self.clock_flg = 1
+
+    def bjorklund(self, steps, pulses):
+        steps = int(steps)
+        pulses = int(pulses)
+        
+        if pulses > steps:
+            raise ValueError    
```

Comment:
It's good practice to add a message to an error in order to make it easier to debug when it occurs. For example:

raise ValueError(f"pulses ({pulses}) unexpectedly larger than steps ({steps})")

Reply:
I would suggest avoiding throwing an exception, which terminates the script, and instead use the [clamp](https://allen-synthesis.github.io/EuroPi/generated/europi.html#europi.clamp) function. Something like this:

```python
pulses = clamp(int(pulses), 0, steps)
```

Reply:
This code block is directly copied from the python implementation of the Bjorklund algorithm repository. I simply delete this part as the input of this function is already bounded by other logic.